scarb/src/internal/offloader.rs

```diff
@@ -0,0 +1,74 @@
+use crate::core::Workspace;
+use anyhow::{Context, Result};
+use std::marker::PhantomData;
+use std::sync::mpsc;
+use std::thread::Scope;
+use std::{mem, thread};
+use tracing::debug;
+
+pub type BoxedWorkload<'a> = Box<dyn FnOnce(&'a Workspace<'a>) -> Result<()> + Send>;
+pub type OffloaderSink<'a> = mpsc::Sender<(&'static str, BoxedWorkload<'a>)>;
+pub type OffloaderStream<'a> = mpsc::Receiver<(&'static str, BoxedWorkload<'a>)>;
+
+pub struct Offloader<'a> {
+    handle: Option<thread::ScopedJoinHandle<'a, Result<()>>>,
+    _phantom_data: PhantomData<Workspace<'a>>,
+    sink: OffloaderSink<'a>,
+}
+
+impl<'a> Offloader<'a> {
+    pub fn new(scope: &'a Scope<'a, '_>, ws: &'a Workspace<'a>) -> Self {
+        let ws = unsafe {
+            // This should be safe, as we know we will join the artifact writer thread before
+            // the workspace is dropped.
+            mem::transmute::<&Workspace<'_>, &Workspace<'_>>(ws)
```

Comment:
clippy will probably yell at this, but it'd be good to explicilyt name the lifetimes here